requests-for-comments/2023-11-22-scheduled-flag-changes.md

```diff
@@ -0,0 +1,84 @@
+
+## Solution
+
+Scheduling feature flag changes allows users to automatically change flag properties at a future point in time. For example, a user can schedule an initial release of the flag for internal testing, then enable it to the beta users group 2 days later, and finally roll it out to 100% of the users in 2 weeks.
+
+## Scope
+
+This feature should allow the following scheduled modifications to a feature flag:
+
+-   **Enable/disable the flag** (field: `is_active`)
+-   **Add rollout conditions to the flag** (field: `filters`)
+
+We want to be able to _generalize_ this implementation to other models such as `Survey` or `Experiment`.
+
+## Implementation
+
+To schedule future changes, we can have a scheduling API, specifying the model the operate on, the id, the changes to be executed, and the schedule time.
+
+```
+POST /api/projects/:project_id/schedule
+{
+    "model": "FeatureFlag",
+    "id": 57,
+    "changes": [ ... ]
+    "scheduled_at": <timestamp>
+}
+```
+
+A change to the record can be expressed as a JSON, specifying the field to be modified and the new value to be set:
+
+**Example 1: Enable the flag**
+
+```
+{
+    "field": "is_active",
+    "value": "false"
+}
+```
+
+**Example 2: Add a rollout condition to the flag**
+
+```
+{
+    "field": “filter”,
+    "value": {
+        "variant": null,
+        "properties": [{
+            "key": "$geoip_country_name",
+            "type": "person",
+            "value": ["Australia"],
+            "operator": "exact"
+        }],
+        "rollout_percentage": 20
+    }
+}
+```
+
+_At_ or _after_ the time specified, the change needs to be propagated onto the record.
+
+We already use Celery with a Redis queue to execute various background tasks. However, because the Redis queue is not durable, we cannot schedule tasks at the precise point in the future. A possible solution here is to have Postgres table of scheduled changes. Celery could periodically poll this table, e.g. every minute. During each poll, Celery would process any rows where the _scheduled_at_ time has already passed.
+
+The `scheduled_flag_changes` table would have the following schema:
+
+| team_id | id  | changes (JSON) | scheduled_at (timestamp) | status (enum) |
+| ------- | --- | -------------- | ------------------------ | ------------- |
+| 1       | 57  | [ ... ]        | 2023-11-23 15:30:00      | scheduled     |
+
+The table would be indexed on the `scheduled_at` column to efficiently retrieve the columns that have the scheduled\*at value in the past. For every row retrieved, the task would **atomically**:
+
+-   Modify the field on the feature flag record
+-   Mark the row as `completed`
+
+We would need to have separate tables for other models such as `scheduled_survey_changes`, due to the schema differences and possibly different scaling requirements.
```

Comment:
Agreed. Both celery tasks can reuse a lot of the queuing logic though

Reply:
I'm not clear why we'll need separate tables for each model, why can't / shouldn't they be all in one model?

Scaling imo is not a concern here. Given we have [not that many](http://metabase-prod-us/question#eyJkYXRhc2V0X3F1ZXJ5Ijp7InF1ZXJ5Ijp7InNvdXJjZS10YWJsZSI6MTQ4LCJhZ2dyZWdhdGlvbiI6W1siY291bnQiXV19LCJ0eXBlIjoicXVlcnkiLCJkYXRhYmFzZSI6MzR9LCJkaXNwbGF5IjoidGFibGUiLCJ2aXN1YWxpemF0aW9uX3NldHRpbmdzIjp7fX0=) flags, assuming 1 thing per flag, all scheduled at the same time, that's 20k items to operate on today, and assuming 10x growth, 200k in the next 2 years. [and flags are biggest w.r.t team feature success right now].

Since we're inputting arbitrary json for the operation to do, schema differences also don't mean much to us, the models themselves determine how to consume the json and do the operation, while the scheduler just has the scheduler logic + model + payload (json).

Reply:
I guess the one disadvantage is not having foreign keys / cascade deletes functionality via the db if it's all in one table, but reasonable trade-off, given we can easily handle deleted flags / any inconsistencies.

Although, worth considering if we want a failure status at all? I can't yet think of how we'll end up in such a state, given other constraints, so maybe ok to defer till we find such a case

Reply:
no, don't think so. 👍 

---

Although, failure reason does make more sense now, thinking about this, because say for example the db update is invalid (does not pass validation filters for example) - we want to tell the user the scheduled update failed for whatever reason.

While the frontend should make sure users can't input invalid stuff, API level updates can bypass this, so our core validation should be on the backend